drivers/uart: stm32: Fix pm_constraint handling

erwango opened the pull request:
Introduce new logic to set/release pm_constraint during serial TX
transactions.

First change is to introduce an internal flag and utility functions
to control the set/release constraint balancing per uart device.
This way, whatever the mix of transactions or API calls, we
ensure a single uart device can only do 1 or 0 to the PM state
constraint. Constraint can't then be set more than once, released w/o
having been set or released more than it was set.

The last part of the change reworks the triggers for constraints
set/release operations.
In order not to disturb driver operations, if irq driven mode or PM is
enabled, don't enable TC interrupt handling by default.
Instead, map the pm_constraint setting to the way TC flag is handled
in normal mode of operations (irq driven or async).
As a consequence, in irq driven mode, pm_constraint is set/released on
tx_enable/tx_disable api calls, which gives API user full control
on transaction protection vs low power operations.
Finally, we emulate the same behavior on TX poll transaction, by
enabling TC irq at the start of a stream and disabling TC irq once
stream is completed. This is controlled with a dedicated device flag.


Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango committed:
drivers/uart: stm32: Fix pm_constraint handling

Introduce new logic to set/release pm_constraint during serial TX
transactions.

First change is to introduce an internal flag and utility functions
to control the set/release constraint balancing per uart device.
This way, whatever the mix of transactions or API calls, we
ensure a single uart device can only do 1 or 0 to the PM state
constraint. Constraint can't then be set more than once, released w/o
having been set or released more than it was set.

The last part of the change reworks the triggers for constraints
set/release operations.
In order not to disturb driver operations, if irq driven mode or PM is
enabled, don't enable TC interrupt handling by default.
Instead, map the pm_constraint setting to the way TC flag is handled
in normal mode of operations (irq driven or async).
As a consequence, in irq driven mode, pm_constraint is set/released on
tx_enable/tx_disable api calls, which gives API user full control
on transaction protection vs low power operations.
Finally, we emulate the same behavior on TX poll transaction, by
enabling TC irq at the start of a stream and disabling TC irq once
stream is completed. This is controlled with a dedicated device flag.


Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
@ceolin, @gmarull This is an attempt to solve issues seen with https://github.com/zephyrproject-rtos/zephyr/pull/38432 (#38521, #38519, and others not loggued). I'd welcome your feedbacks and suggestions

ycsin commented on drivers/serial/uart_stm32.c:
```diff
@@ -53,6 +53,33 @@ LOG_MODULE_REGISTER(uart_stm32);
+}
+
+static void uart_stm32_pm_constraint_release(const struct device *dev,
+					     enum pm_state state)
```
Is `state` a typo?

erwango replied:
Yes, just fixed it

ycsin reviewed (approved):
I have just tested the latest changes, UARTs work fine, no garbage data, MCU able to enter LPM:

Don't mind the build time, cause `main.c` isn't changed:
```
I: Starting bootloader
I: Primary image: magic=good, swap_type=0x1, copy_done=0x3, image_ok=0x1
I: Scratch: magic=unset, swap_type=0x1, copy_done=0x3, image_ok=0x3
I: Boot source: primary slot
I: Swap type: none
I: Bootloader chainload address offset: 0x10000
I: Jumping to the first image slot


gmoc:~ $ ***** delaying boot 2000ms (per build configuration) *****


Build time: Sep 21 2021 12:19:59 Malaysia!


[00:00:04.01Odometer value: 2.5220561 km
4,000] <inf> telemetry_mqtt: Init OK
[00:00:04.015,000] <inf> telemetry: Init OK
[00:00:04.515,000] <inf> timesync: Initialized
```

STM32G0B1RE + EC21 `gsm_ppp` modem with uart_mux, `CONFIG_PM`, `CONFIG_PM_DEVICE` enabled.
Thanks! @erwango

ycsin commented on drivers/serial/uart_stm32.c:
```diff
@@ -456,8 +483,19 @@ static void uart_stm32_poll_out(const struct device *dev,
+	if (!data->tx_poll_stream_on) {
+		data->tx_poll_stream_on = true;
+
+		/* do not allow system to suspend until stream transmission has completed */
```
A bit nitpicking, but maybe capitalize this while you are here?

ycsin commented:
Sorry, I might have discover another bug:

Something caused the log subsys to stop working I believe, `printf` still works though, I'm still finding out why

bbilas commented:
Seems to work pretty well on stm32f4_disco with an external gsm module.

erwango committed:
drivers/uart: stm32: Simplify code around pm_constraints handling

Now that we're clearer around pm constraints management in various
TX cases (poll streams, irq driven or async), make some code
simplifications to ease readability.

Signed-off-by: Erwan Gouriou <erwan.gouriou@linaro.org>

erwango commented:
> Sorry, I might have discover another bug:

@ycsin Would you be able to provide conditions for reproduction?
I'm trying to reproduce the issue using `samples/subsys/logging/logger`, with `CONFIG_PM=y`, adding various sleeps bit so far I can't find any issue.

ycsin commented:
> > Sorry, I might have discover another bug:
> 
> @ycsin Would you be able to provide conditions for reproduction?
> I'm trying to reproduce the issue using `samples/subsys/logging/logger`, with `CONFIG_PM=y`, adding various sleeps bit so far I can't find any issue.

What I know in my application is that I have threads doing normal `LOG_*` and this 'location' thread doing both `LOG_*` and `printf` when it wants to print floating point, when the GNSS starts outputting NMEA sentences.

- If the GNSS didn't get a fix and didn't output NMEA, and the application go to sleep, the log is still working after that.
- But if the GNSS gets a fix, `printf` the data and go to sleep immediately, then log is dead after that.

I might have to setup something simpler on a nucleo board to test

erwango commented:
> But if the GNSS gets a fix, printf the data and go to sleep immediately, then log is dead after that.

Is that only the LOG and printf will work correctly ?
Or there is no more output at all from the serial device ?

ycsin commented:
> > But if the GNSS gets a fix, printf the data and go to sleep immediately, then log is dead after that.
> 
> Is that only the LOG and printf will work correctly ?
> Or there is no more output at all from the serial device ?

As far as I can tell the uart is still working.

UART1 that connected with the modem works, I can see it setting up the modem and sending data through internet. From my online dashboard I can see new data being sent.

UART5 that connected with GNSS works, the NMEA sentences are still being parsed.

UART2 that is used by the `shell-uart` & `console` "works", data from `printf` are bring displayed properly.

But no more log msgs, and typing in the shell has no response so I'd suspect that only the LOG and/or SHELL subsys stopped working for some reason.

FRASTM commented on drivers/serial/uart_stm32.c:
```diff
@@ -456,8 +478,23 @@ static void uart_stm32_poll_out(const struct device *dev,
+		 * transmission has completed
+		 */
+		uart_stm32_pm_constraint_set(dev);
+
```
does it save some code with   `uart_stm32_pm_constraint_set(data)

erwango replied:
@FRASTM, I don't get this comment

FRASTM replied:
writing 
`static void uart_stm32_pm_constraint_set(const struct uart_stm32_data *data);`
instead of 
`static void uart_stm32_pm_constraint_set(const struct device *dev);`

erwango replied:
Both cases use a pointer, I'm not sure this would make a difference.
And then we'd need to extract `data` in each callers, which is not done yet.

erwango commented:
@ycsin For the shell, with PM enabled, did you configure UART so it could wake up the system on RX?
Also, can you point to a SHA1 where the full system was working ?

FRASTM commented on drivers/serial/uart_stm32.c:
```diff
@@ -799,16 +839,30 @@ static void uart_stm32_dma_rx_flush(const struct device *dev)
 static void uart_stm32_isr(const struct device *dev)
 {
 	struct uart_stm32_data *data = DEV_DATA(dev);
+#if defined(CONFIG_PM) || defined(CONFIG_UART_ASYNC_API)
```
below, there is also a `uart_stm32_isr() `for the CONFIG_PM

erwango replied:
The other `uart_stm32_isr` is specific for !CONFIG_UART_ASYNC_API && ! IRQ_DRIVEN

FRASTM replied:
yes, removed in the next commit

ycsin commented:
> @ycsin For the shell, with PM enabled, did you configure UART so it could wake up the system on RX?

Is this supported in the current `uart_stm32`?

> Also, can you point to a SHA1 where the full system was working ?

The `uart_stm32.c` version before #38432, in terms of SHA1, 88b60f31c74aa2a35d4869bc4ec463c52fc1f451 should be good.

What I currently have is something close to the tip of main branch, then manually revert #38432 based on the diff

erwango commented:
> Is this supported in the current uart_stm32, for me currently after the MCU enters deep sleep the UART isn't responsive anymore?

That's not supported actually.
So when PM is enabled, in order to be able to input to the shell, the system should be woken up from other another source.

erwango commented:
@ycsin I've been testing this patch on `samples/subsys/logging/logger,` adding CONFIG_SHELL and CONFIG_PM and few twists to increase the use cases, I've not faced any issue so far (other than not  being able to input in shell while system is suspended, which i expected).
Given that current PR solves a number of issues and unlock most users, would you be ok to merge current version?
Then you can raise a dedicated issue for your problem  (hopefully with a reproducible set up), and we can start again from there.

ycsin commented:
> @ycsin I've been testing this patch on `samples/subsys/logging/logger,` adding CONFIG_SHELL and CONFIG_PM and few twists to increase the use cases, I've not faced any issue so far (other than not being able to input in shell while system is suspended, which i expected).
> Given that current PR solves a number of issues and unlock most users, would you be ok to merge current version?
> Then you can raise a dedicated issue for your problem (hopefully with a reproducible set up), and we can start again from there.

Yup no problem, please go ahead. I will try it when I rebase my branch later. Thanks!

bbilas commented:
@erwango Have you tried to use e.g usart0 as shell/log output and  enable hw flow control in usart2? I just used this configuration and usart0 isn't working at all.
